### PR TITLE
CI: migrate travis to circle

### PR DESCRIPTION
**Pull request by alvin-huang**
This PR migrates Terraform from Travis over to CircleCI. I tried to copy most of the configuration and added in bits that I thought would be useful from some of the other projects I have worked on. 

A summary of the more notable changes:
1. I am bumping consul to 1.7.2 (latest) to run tests against.
2. There is a check to make sure `make generate` does not generate a dirty git directory. This should be a command that is run locally and committed back to the repo.
3. Go tests are [parallelized](https://app.circleci.com/pipelines/github/hashicorp/terraform/20/workflows/0a188038-4f96-4125-948d-15ac865072c4) across 4 executors based on package runtimes calculated by `go list ./... | circleci tests split --split-by=timings --timings-type=classname`. They run with the `-cover` flag and these coverage results are aggregated together and uploaded to Codecov. Example: https://codecov.io/gh/hashicorp/terraform/commit/4e52a7a878c5550fcf218d756f9c49b4b17f869e. More info on test splitting can be found in the [CircleCI docs](https://circleci.com/docs/2.0/parallelism-faster-jobs/). 
4. There are[ parallel builds](https://app.circleci.com/pipelines/github/hashicorp/terraform/20/workflows/8616ec7d-10de-4606-8097-f06fb6c4ae82) of all the various architectures Terraform builds that runs on every PR. These artifacts and test builds can be directly downloaded from the CircleCI interface. For example: https://app.circleci.com/pipelines/github/hashicorp/terraform/20/workflows/8616ec7d-10de-4606-8097-f06fb6c4ae82/jobs/85/artifacts. This was something we ran into on Consul where there was an upstream dependency that had a build tag to exclude Solaris that we couldn't build since Consul builds on Solaris. Rather than finding this out at release time, these parallel builds will give you a quick smoke test that everything still compiles for all the OS_ARCH that Terraform builds. 
5. The tests are no longer using the `make test` and `make e2etest` make targets because they use gotestsum to generate XML that CircleCI can then understand and parse for failed test results.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,17 +1,175 @@
...
+      - run:
+          name: install consul
+          command: |
+            curl -sLo consul.zip https://releases.hashicorp.com/consul/${CONSUL_VERSION}/consul_${CONSUL_VERSION}_linux_amd64.zip
+            unzip consul.zip
+            mkdir -p ~/bin
+            mv consul ~/bin
+            echo 'export PATH="~/bin:$PATH"'
+      - run: mkdir -p $TEST_RESULTS_DIR
+      - run: |
```
Comment (apparentlymart) [inline .circleci/config.yml]: I noticed this one (and its equivalent under `go-test-e2e`) generate quite a clunky heading in the CircleCI UI.



Looking at some of the other examples in this file it looks like we can set `name` to customize caption on those. Could we do that here to set it to something shorter like "Running Tests"? :thinking:

**Comment by alvin-huang** [comment]
Ah that is probably because [this PR](https://github.com/hashicorp/terraform/pull/24483) was not rebased into the branches for the PR. CircleCI runs the config in the specific branch rather than run the master branch config for all branches.

File: .circleci/config.yml (hunk 1 above)
Comment (alvin-huang) [reply .circleci/config.yml]: I added the naming per your request!

**Comment by apparentlymart** [comment]
Thanks! This seems to be all working, per the reports on this PR, so I'm going to go ahead and merge this so we can rebase or cherry-pick it onto other branches as appropriate.

**Pull request merged**